compiler/optimizer/VPHandlers.cpp

```diff
@@ -1151,31 +1151,86 @@ TR::Node *constrainBCDAggrLoad(OMR::ValuePropagation *vp, TR::Node *node)
+               {
+               TR::VPConstString *constString = baseVPConstraint->getClassType()->asConstString();
 
-            uint16_t ch = constString->charAt(((TR::Compiler->target.is64Bit() ? index->getLongIntLow() : index->getInt())
-                                               - TR::Compiler->om.contiguousArrayHeaderSizeInBytes()) / 2, vp->comp());
-            if (ch != 0)
+               uint16_t ch = constString->charAt(((TR::Compiler->target.is64Bit() ? index->getLongIntLow() : index->getInt())
+                                                  - TR::Compiler->om.contiguousArrayHeaderSizeInBytes()) / 2, vp->comp());
+               if (ch != 0)
+                  {
+                  vp->replaceByConstant(node, TR::VPShortConst::create(vp, ch), isGlobal);
+                  return node;
+                  }
+               }
+            }
+         TR::VPConstraint *arrayVPConstraint = vp->getConstraint(array, isGlobal);
+         if (arrayVPConstraint)
+            {
+#ifdef J9_PROJECT_SPECIFIC
+            TR::KnownObjectTable *knot = vp->comp()->getKnownObjectTable();
+            TR::VPKnownObject *kobj = arrayVPConstraint->getKnownObject();
+            if (knot && kobj)
                {
-               vp->replaceByConstant(node, TR::VPIntConst::create(vp, ch), true);
-               return node;
+               TR::KnownObjectTable::Index idx = kobj->getIndex();
+               if (kobj->isArrayWithConstantElements(vp->comp()) && kobj->isPrimitiveArray(vp->comp()))
+                  {
+                  TR::VMAccessCriticalSection constrainAnyIntLoadCriticalSection(vp->comp(),
+                     TR::VMAccessCriticalSection::tryToAcquireVMAccess);
+                  if (constrainAnyIntLoadCriticalSection.hasVMAccess())
+                     {
+                     uintptrj_t arrayObj = knot->getPointer(idx);
+                     uintptrj_t offset = TR::Compiler->target.is64Bit()
+                                         ? (uintptrj_t)index->getUnsignedLongInt() : (uintptrj_t)index->getUnsignedInt();
+                     uintptrj_t lengthInElements = TR::Compiler->om.getArrayLengthInElements(vp->comp(), arrayObj);
+                     TR::DataType elementType = kobj->getPrimitiveArrayDataType();
+                     uintptrj_t lengthInBytes = lengthInElements * TR::DataType::getSize(elementType);
+                     uintptrj_t validStart = TR::Compiler->om.contiguousArrayHeaderSizeInBytes();
+                     uintptrj_t validEnd = validStart + lengthInBytes;
+                     uintptrj_t accessSize = TR::DataType::getSize(dataType);
+                     uintptrj_t accessEnd = offset + accessSize;
+                     if (validStart <= offset && offset < accessEnd && accessEnd <= validEnd)
+                        {
+                        uintptrj_t elementAddress = TR::Compiler->om.getAddressOfElement(arrayObj, offset);
+                        switch (obj->getPrimitiveArrayDataType())
```

Comment:
This should be switching on the type of the load (`dataType`)

Reply:
fixed